core/payments/ondemand/on_demand_payment_validator.go

```diff
@@ -0,0 +1,179 @@
+// NewOnDemandPaymentValidator creates a new OnDemandPaymentValidator with specified cache size
+func NewOnDemandPaymentValidator(
+	logger logging.Logger,
+	// the maximum number of OnDemandLedger entries to be kept in the LRU cache
+	maxLedgers int,
+	// expected to be initialized and have its background update thread started
+	onChainState meterer.OnchainPayment,
+	dynamoClient *dynamodb.Client,
+	// the name of the dynamo table where on-demand payment information is stored
+	onDemandTableName string,
+) (*OnDemandPaymentValidator, error) {
+	if onChainState == nil {
+		return nil, errors.New("onChainState cannot be nil")
+	}
+	cache, err := lru.NewWithEvict(
+		maxLedgers,
+		func(accountAddress gethcommon.Address, _ *OnDemandLedger) {
+			logger.Infof("evicted account %s from LRU on-demand ledger cache", accountAddress.Hex())
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("new LRU cache with evict: %w", err)
+	}
+
+	if dynamoClient == nil {
+		return nil, errors.New("dynamo client cannot be nil")
+	}
+
+	if onDemandTableName == "" {
+		return nil, errors.New("on demand table name cannot be empty")
+	}
+
+	return &OnDemandPaymentValidator{
+		logger:            logger,
+		ledgers:           cache,
+		onChainState:      onChainState,
+		dynamoClient:      dynamoClient,
+		onDemandTableName: onDemandTableName,
+	}, nil
+}
+
+// Debit validates an on-demand payment for a blob dispersal
+// The caller is responsible for verifying the signature before calling this method
+func (pv *OnDemandPaymentValidator) Debit(
+	ctx context.Context,
+	accountID gethcommon.Address,
+	symbolCount uint32,
+	quorumNumbers []uint8,
+) error {
+	ledger, err := pv.getOrCreateLedger(ctx, accountID)
+	if err != nil {
+		return fmt.Errorf("get or create on-demand ledger: %w", err)
+	}
+
+	_, err = ledger.Debit(ctx, symbolCount, quorumNumbers)
+	if err != nil {
+		return fmt.Errorf("debit on-demand payment: %w", err)
+	}
+
+	return nil
+}
+
+// getOrCreateLedger gets an existing on-demand ledger from the cache, or creates a new one if it doesn't exist
+func (pv *OnDemandPaymentValidator) getOrCreateLedger(
+	ctx context.Context,
+	accountID gethcommon.Address,
+) (*OnDemandLedger, error) {
+	// Fast path: check if ledger already exists in cache
+	if ledger, exists := pv.ledgers.Get(accountID); exists {
+		return ledger, nil
+	}
+
+	// Slow path: acquire lock and check again
+	pv.ledgerCreationLock.Lock()
+	defer pv.ledgerCreationLock.Unlock()
+
+	if ledger, exists := pv.ledgers.Get(accountID); exists {
+		return ledger, nil
+	}
+
+	onDemandPayment, err := pv.onChainState.GetOnDemandPaymentByAccount(ctx, accountID)
+	if err != nil {
+		return nil, fmt.Errorf("get on-demand payment for account %v: %w", accountID.Hex(), err)
+	}
+
+	pricePerSymbol := big.NewInt(int64(pv.onChainState.GetPricePerSymbol()))
+	minNumSymbols := pv.onChainState.GetMinNumSymbols()
+
+	cumulativePaymentStore, err := NewCumulativePaymentStore(pv.dynamoClient, pv.onDemandTableName, accountID)
```

Comment:
bike-shed: wondering if we should consider adding a TTL to the LRU to force any cached ledger to refetch these items in case they have changed onchain?

Reply:
There will be consideration for changing vault values in an upcoming PR.

MVP will use a periodic fetch of vault data, matching the current pattern in use. A fast follow will convert this to an event subscription model.

In both cases, evicting from the cache isn't necessary: we'll just update the parameters in-place